File: src/web/routes/adjustments.py
import getpass
import sqlite3

from fastapi import APIRouter, Depends, Form, Request
from fastapi.responses import RedirectResponse

from src.db.repositories import adjustment_repo, entity_repo, group_coa_repo, period_repo
from src.db.repositories.adjustment_repo import VALID_TYPES
from src.engines.adjustment_engine import create_adjustment, copy_prior_month
from src.web.deps import get_db, templates

router = APIRouter()


@router.get("/periods/{period_str}/adjustments")
def list_adjustments(request: Request, period_str: str, conn: sqlite3.Connection = Depends(get_db)):
    period = period_repo.get_by_str(conn, period_str)
    if period is None:
        return RedirectResponse(f"/?flash=No such period&flash_kind=error", status_code=303)

    adjustments = []
    for adj in adjustment_repo.list_for_period(conn, period["period_id"]):
        lines = adjustment_repo.get_lines(conn, adj["adjustment_id"])
        total_dr = sum(l["debit_amount"] or 0 for l in lines)
        line_summaries = []
        for l in lines:
            cat = group_coa_repo.get_by_id(conn, l["group_account_id"])
            entity_code = ""
            if l["entity_id"]:
                e = conn.execute("SELECT entity_code FROM entities WHERE entity_id = ?",
                                  (l["entity_id"],)).fetchone()
                entity_code = e["entity_code"] if e else ""
            line_summaries.append({
                "entity": entity_code or "(group)",
                "category": cat["account_name"] if cat else "?",
                "debit": l["debit_amount"], "credit": l["credit_amount"],
            })
        adjustments.append({
            "ref": adj["external_ref"], "type": adj["adjustment_type"], "narration": adj["narration"],
            "status": adj["status"], "total": total_dr, "lines": line_summaries,
        })

    prior = period_repo.get_prior(conn, period["period_id"])
    return templates.TemplateResponse(request, "adjustments_list.html", {
        "request": request, "period_str": period_str, "adjustments": adjustments,
        "period_status": period["status"], "has_prior": prior is not None,
    })


@router.get("/periods/{period_str}/adjustments/new")
def new_adjustment_form(request: Request, period_str: str, conn: sqlite3.Connection = Depends(get_db)):
    entities = entity_repo.list_all(conn)
    categories = group_coa_repo.list_leaf_categories(conn)
    return templates.TemplateResponse(request, "adjustments_form.html", {
        "request": request, "period_str": period_str, "entities": entities,
        "categories": [c["account_name"] for c in categories], "types": sorted(VALID_TYPES),
        "errors": [], "form": {},
    })


@router.post("/periods/{period_str}/adjustments/new")
def create_adjustment_route(request: Request, period_str: str,
                              external_ref: str = Form(...), adjustment_type: str = Form(...),
                              narration: str = Form(...),
                              entity_code: list[str] = Form(default=[]),
                              category: list[str] = Form(default=[]),
                              debit: list[str] = Form(default=[]),
                              credit: list[str] = Form(default=[]),
                              conn: sqlite3.Connection = Depends(get_db)):
    period = period_repo.get_by_str(conn, period_str)
    if period is None:
        return RedirectResponse(f"/?flash=No such period&flash_kind=error", status_code=303)

    entities = entity_repo.list_all(conn)
    categories = group_coa_repo.list_leaf_categories(conn)
    errors = []
    lines = []
    for i, (ecode, cat_name, dr, cr) in enumerate(zip(entity_code, category, debit, credit), start=1):
        if not cat_name:
            continue
        cat = group_coa_repo.get_by_name(conn, cat_name)
        if cat is None:
            errors.append(f"Line {i}: unknown category '{cat_name}'")
            continue
        entity_id = None
        if ecode:
            e = entity_repo.get_by_code(conn, ecode)
            if e is None:
                errors.append(f"Line {i}: unknown entity '{ecode}'")
                continue
            entity_id = e["entity_id"]
        lines.append({
            "entity_id": entity_id, "group_account_id": cat["group_account_id"],
            "debit_amount": float(dr or 0), "credit_amount": float(cr or 0),
        })

    if not errors:
        try:
            create_adjustment(conn, period["period_id"], external_ref, adjustment_type, narration,
                                lines, user=getpass.getuser())
            conn.commit()
        except ValueError as e:
            errors = str(e).split("\n")

    if errors:
        return templates.TemplateResponse(request, "adjustments_form.html", {
            "request": request, "period_str": period_str, "entities": entities,
            "categories": [c["account_name"] for c in categories], "types": sorted(VALID_TYPES),
            "errors": errors,
            "form": {"external_ref": external_ref, "adjustment_type": adjustment_type, "narration": narration},
        }, status_code=400)

    return RedirectResponse(
        f"/periods/{period_str}/adjustments?flash=Created {external_ref} as draft&flash_kind=success",
        status_code=303)


@router.post("/periods/{period_str}/adjustments/apply-all")
def apply_all_route(period_str: str, conn: sqlite3.Connection = Depends(get_db)):
    period = period_repo.get_by_str(conn, period_str)
    try:
        n = adjustment_repo.apply_all(conn, period["period_id"], user=getpass.getuser())
        conn.commit()
        msg, kind = f"Applied {n} draft adjustment(s) — re-run Consolidate to see the effect", "success"
    except ValueError as e:
        msg, kind = str(e), "error"
    return RedirectResponse(f"/periods/{period_str}/adjustments?flash={msg}&flash_kind={kind}", status_code=303)


@router.post("/periods/{period_str}/adjustments/{ref}/delete")
def delete_adjustment_route(period_str: str, ref: str, conn: sqlite3.Connection = Depends(get_db)):
    period = period_repo.get_by_str(conn, period_str)
    row = conn.execute(
        "SELECT adjustment_id FROM adjustments WHERE period_id = ? AND external_ref = ?",
        (period["period_id"], ref),
    ).fetchone()
    if row:
        try:
            adjustment_repo.delete(conn, period["period_id"], row["adjustment_id"], user=getpass.getuser())
            conn.commit()
            msg, kind = f"Deleted {ref}", "success"
        except ValueError as e:
            msg, kind = str(e), "error"
    else:
        msg, kind = f"No adjustment {ref} found", "error"
    return RedirectResponse(f"/periods/{period_str}/adjustments?flash={msg}&flash_kind={kind}", status_code=303)


@router.post("/periods/{period_str}/adjustments/copy-prior")
def copy_prior_route(period_str: str, conn: sqlite3.Connection = Depends(get_db)):
    period = period_repo.get_by_str(conn, period_str)
    try:
        n = copy_prior_month(conn, period["period_id"], user=getpass.getuser())
        conn.commit()
        msg, kind = f"Copied {n} adjustment(s) from the prior period as drafts", "success"
    except ValueError as e:
        msg, kind = str(e), "error"
    return RedirectResponse(f"/periods/{period_str}/adjustments?flash={msg}&flash_kind={kind}", status_code=303)

File: docs/src/engines/adjustments.js
// Port of src/engines/adjustment_engine.py — structured double-entry adjustments,
// replaces manually editing TB numbers. Debits must equal credits, no line may carry
// both a debit and a credit, at least 2 lines, narration required.
import { persistence } from "../db/persistence.js";
import * as adjustmentRepo from "../db/adjustment-repo.js";
import * as periodRepo from "../db/period-repo.js";
import * as groupCoaRepo from "../db/group-coa-repo.js";
import { log as auditLog } from "../services/audit-service.js";

const TOLERANCE_IDR = 0;

export function validate(narration, lines, adjustmentType = null) {
  const errors = [];
  if (!narration || !narration.trim()) {
    errors.push("Adjustment narration is required");
  }
  if (adjustmentType !== null && !adjustmentRepo.VALID_TYPES.has(adjustmentType)) {
    errors.push(
      `Unknown adjustment type '${adjustmentType}' — must be one of ${Array.from(adjustmentRepo.VALID_TYPES).sort().join(", ")}`
    );
  }
  if (lines.length < 2) {
    errors.push("An adjustment needs at least 2 lines");
  }
  const totalDr = lines.reduce((s, l) => s + (l.debitAmount || 0.0), 0.0);
  const totalCr = lines.reduce((s, l) => s + (l.creditAmount || 0.0), 0.0);
  if (Math.abs(totalDr - totalCr) > TOLERANCE_IDR) {
    errors.push(`Adjustment does not balance: debits=${totalDr.toFixed(2)} credits=${totalCr.toFixed(2)}`);
  }
  lines.forEach((l, i) => {
    if ((l.debitAmount || 0.0) && (l.creditAmount || 0.0)) {
      errors.push(`Line ${i + 1}: cannot carry both a debit and a credit`);
    }
  });
  return errors;
}

export function createAdjustment(periodId, externalRef, adjustmentType, narration, lines, user = null) {
  const period = persistence.get("SELECT * FROM periods WHERE period_id = ?", [periodId]);
  if (period.status === "locked") {
    throw new Error(`Period ${period.year}-${String(period.month).padStart(2, "0")} is locked; cannot add adjustments`);
  }

  const errors = validate(narration, lines, adjustmentType);
  if (errors.length) {
    throw new Error(`Adjustment '${externalRef}' is invalid:\n` + errors.join("\n"));
  }

  return adjustmentRepo.createOrReplace(periodId, externalRef, adjustmentType, narration, lines, user);
}

// Resolves entity codes and category names so the audit log's before/after JSON reads
// like a human-reviewable journal entry, not raw foreign keys.
function snapshot(adjustmentRow, lines) {
  const lineSnapshots = lines.map((l) => {
    let entityCode = null;
    if (l.entity_id) {
      const e = persistence.get("SELECT entity_code FROM entities WHERE entity_id = ?", [l.entity_id]);
      entityCode = e ? e.entity_code : null;
    }
    const category = groupCoaRepo.getById(l.group_account_id);
    return {
      entity: entityCode,
      category: category ? category.account_name : null,
      debit: l.debit_amount,
      credit: l.credit_amount,
    };
  });
  return {
    externalRef: adjustmentRow.external_ref,
    type: adjustmentRow.adjustment_type,
    narration: adjustmentRow.narration,
    status: adjustmentRow.status,
    lines: lineSnapshots,
  };
}

// Edit an existing adjustment (draft or applied) in place. Validates, requires the period
// open, snapshots before/after into the audit log, and resets status to draft.
//
// externalRef is immutable here by design — it's always read from the existing row, never
// taken from the caller, so an edit can't collide with the (period_id, external_ref)
// upsert key or become indistinguishable from creating a second adjustment.
export function updateAdjustment(periodId, adjustmentId, adjustmentType, narration, lines, user = null) {
  periodRepo.requireOpen(periodId, "edit an adjustment");

  const existing = persistence.get("SELECT * FROM adjustments WHERE adjustment_id = ? AND period_id = ?", [adjustmentId, periodId]);
  if (!existing) {
    throw new Error(`Adjustment ${adjustmentId} not found in this period`);
  }

  const before = snapshot(existing, adjustmentRepo.getLines(adjustmentId));

  const errors = validate(narration, lines, adjustmentType);
  if (errors.length) {
    throw new Error(`Adjustment '${existing.external_ref}' is invalid:\n` + errors.join("\n"));
  }

  adjustmentRepo.createOrReplace(periodId, existing.external_ref, adjustmentType, narration, lines, user);

  const afterRow = persistence.get("SELECT * FROM adjustments WHERE adjustment_id = ?", [adjustmentId]);
  const after = snapshot(afterRow, adjustmentRepo.getLines(adjustmentId));

  auditLog("edit", "adjustments", String(adjustmentId), { oldValue: JSON.stringify(before), newValue: JSON.stringify(after), user });

  return adjustmentId;
}

// Copies every adjustment from the chronologically prior period into the target period as
// new drafts, narration tagged "(copied from [period])" — matching the user's existing
// copy-prior-month habit.
export function copyPriorMonth(targetPeriodId, user = null) {
  periodRepo.requireOpen(targetPeriodId, "copy adjustments into it");

  const prior = periodRepo.getPrior(targetPeriodId);
  if (!prior) {
    throw new Error("No prior period exists to copy adjustments from");
  }

  const sourceAdjustments = adjustmentRepo.listForPeriod(prior.period_id);
  let copied = 0;
  for (const adj of sourceAdjustments) {
    const lines = adjustmentRepo.getLines(adj.adjustment_id).map((l) => ({
      entityId: l.entity_id,
      groupAccountId: l.group_account_id,
      debitAmount: l.debit_amount,
      creditAmount: l.credit_amount,
      lineNarration: l.line_narration,
    }));
    let narration = adj.narration;
    const tag = `(copied from ${prior.year}-${String(prior.month).padStart(2, "0")})`;
    if (!narration.includes(tag)) narration = `${narration} ${tag}`;
    adjustmentRepo.createOrReplace(targetPeriodId, adj.external_ref, adj.adjustment_type, narration, lines, user, prior.period_id);
    copied++;
  }
  return copied;
}
